gustaf/ffd.py

```diff
@@ -1,4 +1,571 @@
+        )
+        self._logd("FFD successful.")
+        self.calculated = True
+    
+
+    @property
+    def control_points(self):
+        """
+        Returns current spline's control points. Can't and don't use this to
+        directly manipulate control points: it won't update
+
+        Returns
+        --------
+        self._spline.control_points: np.ndarray
+        """
+        return self._spline.control_points
+
+    @control_points.setter
+    def control_points(self,
+                       control_points: Union[List[List[float]], np.ndarray]
+                       ):
+        """
+        Sets control points and deforms mesh.
+
+        Parameters
+        -----------
+        control_points: np.ndarray
+
+        Returns
+        --------
+        None
+        """
+        assert self._spline.control_points.shape == control_points.shape,\
+            "Given control points' shape does not match current ones!"
+
+        self._spline.control_points = control_points.copy()
+        self._logd("Set new control points.")
+
+    def deform_for_given_cp(self, values: np.ndarray, mask=None) -> MESH_TYPES:
+        """
+        Set the new control_point values according to the mask and returns the 
+        deformed mesh.
+
+        #TODO I am not sure if this works since CPs might not always be ndarray
+
+        Parameters
+        -----------
+        values: np.ndarray
+          Values to translate. If it is 1D array, it is applied to all masked
+          control points.
+        mask: list-like
+          (Optional) Default is None. If None, it will be applied to all.
+          It can be bool or int.
+
+        Returns
+        --------
+        deformed_mesh: MESH_TYPES
+        """
+        if mask is None:
+            self.control_points += values
+        else:
+            self.control_points[mask] += values
+
+        return self.mesh
+
+    def elevate_degree(self, *args, **kwargs):
+        """
+        Wrapper for Spline.elevate_degree
+
+        Parameters
+        -----------
+        *args:
+        **kwargs:
+
+        Returns
+        --------
+        None
+        """
+        self._spline.elevate_degree(*args, **kwargs)
+#         self._naive_spline.elevate_degree(*args, **kwargs)
+
+    def insert_knots(self, *args, **kwargs):
+        """
+        Wrapper for Spline.insert_knots
+
+        Parameters
+        -----------
+        *args:
+        **kwargs:
+
+        Returns
+        --------
+        None
+        """
+        self._spline.insert_knots(*args, **kwargs)
+#         self._naive_spline.insert_knots(*args, **kwargs)
+
+    def remove_knots(self, *args, **kwargs):
+        """
+        Wrapper for Spline.remove_knots
+
+        Parameters
+        -----------
+        *args:
+        **kwargs:
+
+        Returns
+        --------
+        None
+        """
+        self._spline.remove_knots(*args, **kwargs)
+#         self._naive_spline.remove_knots(*args, **kwargs)
```

Comment:
```suggestion
```

Reply:
Also, remove knots is not always an exact operation so it could be that the splines geometry changes (although very slightly). As a consequence the is_calculated value should also be set to false here